### PR TITLE
feat: Make `addAlgoliaAgent()` and `clearCache()` optional

### PR DESCRIPTION
**Pull request by francoischalifour**
To simplify the usage of future search clients, a good solution is to make `addAlgoliaAgent()` and `clearCache()` optional. This PR also update the now unnecessary `addAlgoliaAgent` option to fake clients in the tests.

### CODE REVIEW

**Comment by francoischalifour** [comment]
@Haroenv This test should already cover this case: https://github.com/algolia/algoliasearch-helper-js/blob/d12cbda2bb8ebafdf3d5f9e442378d3efb7353ee/test/spec/algoliasearch.helper/client.js#L15-L24

File: test/spec/algoliasearch.helper/instantiate.js [hunk 1]
```diff
@@ -1,14 +0,0 @@
-'use strict';
-
-var test = require('tape');
-var algoliasearchHelper = require('../../../index.js');
-
-test('It should instantiate even with an old client that does not support addAlgoliaAgent', function(t) {
```
Comment (bobylito) [inline test/spec/algoliasearch.helper/instantiate.js]: It should still pass actually :)

File: src/algoliasearch.helper.js [hunk 2]
```diff
@@ -119,8 +119,9 @@ var version = require('./version');
  * just an object containing the properties you need from it.
  */
 function AlgoliaSearchHelper(client, index, options) {
-  if (!client.addAlgoliaAgent) console.log('Please upgrade to the newest version of the JS Client.'); // eslint-disable-line
-  else if (!doesClientAgentContainsHelper(client)) client.addAlgoliaAgent('JS Helper ' + version);
+  if (client.addAlgoliaAgent && !doesClientAgentContainsHelper(client)) {
```
Comment (bobylito) [inline src/algoliasearch.helper.js]: We don't explicitly test for addAlgoliaAgent usage with a client with  and without the method.

File: src/algoliasearch.helper.js [hunk 3]
```diff
@@ -1305,7 +1306,7 @@ AlgoliaSearchHelper.prototype._change = function(newState) {
  * @return {AlgoliaSearchHelper}
  */
 AlgoliaSearchHelper.prototype.clearCache = function() {
-  this.client.clearCache();
+  this.client.clearCache && this.client.clearCache();
```
[the remark this answers is not in the document]
Comment (francoischalifour) [reply src/algoliasearch.helper.js]: I added the tests.

File: src/algoliasearch.helper.js (hunk 2 above)
Comment (francoischalifour) [reply src/algoliasearch.helper.js]: I added the tests.

File: test/spec/algoliasearch.helper/client.js [hunk 4]
```diff
@@ -12,6 +12,56 @@ function makeFakeClient() {
   return client;
 }
 
+test("client without addAlgoliaAgent() doesn't throw on instanciation", function(t) {
```
Comment (Haroenv) [inline test/spec/algoliasearch.helper/client.js]: nit: instantiation

Comment (francoischalifour) [reply test/spec/algoliasearch.helper/client.js]: Oops

File: src/algoliasearch.helper.js (hunk 2 above)
Comment (Haroenv) [reply src/algoliasearch.helper.js]: What about exposing `addAlgoliaAgent` like we do for `clearCache`, that way we can assume in InstantSearch implementations that the method exists and just call it, even if it doesn't on the client

**Pull request merged**